Part 1

# 3.Import your analysis libraries, as well as your new customer data set as a dataframe.

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
## Set path
path = r'/Users/leo/Desktop/Career Foundry/Achievement 4/Instacart Basket Analysis 30:10:25'

In [5]:
## Import Customer Data 
customers = pd.read_csv(os.path.join(path,'02 - data','original data','customers.csv'))

# 4.Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [6]:
# Check the columns 
print(customers.columns)

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')


In [7]:
# Rename columns for consistency (lowercase + underscores)
customers.rename(columns={
    'First Name': 'first_name',
    'Surnam': 'surname',
    'Gender': 'gender',
    'STATE': 'state',
    'Age': 'age',
    'date_joined': 'date_joined',
    'n_dependants': 'n_dependants',
    'fam_status': 'fam_status',
    'income': 'income'
}, inplace=True)

In [8]:
# See Data
customers.head(3)

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568

In [9]:
# Check info
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   first_name    194950 non-null  object
 2   surname       206209 non-null  object
 3   gender        206209 non-null  object
 4   state         206209 non-null  object
 5   age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


# 5.Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

In [10]:
# Check for missing values
customers.isnull().sum()

user_id             0
first_name      11259
surname             0
gender              0
state               0
age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

# First name column has 11,259 blanks. (5.5%).
Since the first name column might be useful at a later stage I have decided to keep it and fill blanks with 'unkown'.

In [11]:
customers['first_name'].fillna('Unknown', inplace=True)

/var/folders/19/p3csqgws1vqgpn31c7f5fcqh0000gn/T/ipykernel_35441/701324329.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  customers['first_name'].fillna('Unknown', inplace=True)


In [12]:
# Check for duplicates
customers.duplicated().sum()

np.int64(0)

In [13]:
# Check mixed data types
customers.dtypes

user_id          int64
first_name      object
surname         object
gender          object
state           object
age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [14]:
# convert date joined to datetime format
customers['date_joined'] = pd.to_datetime(customers['date_joined'])

In [15]:
# Confirm change
customers.dtypes

user_id                  int64
first_name              object
surname                 object
gender                  object
state                   object
age                      int64
date_joined     datetime64[ns]
n_dependants             int64
fam_status              object
income                   int64
dtype: object

In [18]:
# Checking Age column to ensure no outliers like negative numbers or extreme values
customers['age'].describe()

count    206209.000000
mean         49.501646
std          18.480962
min          18.000000
25%          33.000000
50%          49.000000
75%          66.000000
max          81.000000
Name: age, dtype: float64

In [19]:
# Check income for outliers like above
customers['income'].describe()

count    206209.000000
mean      94632.852548
std       42473.786988
min       25903.000000
25%       59874.000000
50%       93547.000000
75%      124244.000000
max      593901.000000
Name: income, dtype: float64

## The income column shows realistic values ranging from about $25,000 to $594,000, with an average of around $94,600. The distribution appears reasonable for a national customer base, so no adjustments were necessary.

In [20]:
print(customers['gender'].value_counts(dropna=False))
print(customers['fam_status'].value_counts(dropna=False))

gender
Male      104067
Female    102142
Name: count, dtype: int64
fam_status
married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: count, dtype: int64


I checked the gender and fam_status columns to make sure there were no missing values or inconsistent entries.
The gender column only contains “Male” and “Female,” showing a balanced and clean distribution.
For fam_status, all categories (“married,” “single,” “divorced/widowed,” and “living with parents and siblings”) are clearly defined and make sense.
There are no missing values or irregular categories, so no further cleaning was needed.

# 6. Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)

In [21]:
## Import Data
ords_prods_merge = pd.read_pickle(os.path.join(path,'02 - data','prepared data','ords_prods_merge_task_4.8_task.pkl'))

# Making changes from practice 4.9 task.

In [24]:
## Changing values in prices columns that are >100 to nan
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [25]:
## check new max of prices
ords_prods_merge['prices'].max()

25.0

Merging data frames

In [26]:
# Check Key column has the same data type
print(ords_prods_merge['user_id'].dtype)
print(customers['user_id'].dtype)

int64
int64


In [27]:
ords_prods_cust = ords_prods_merge.merge(customers, on='user_id', how='left')

# I used a left join because I wanted to keep all the existing Instacart order records and only add customer details where a matching user_id was available. This ensures that no order data is lost during the merge.

In [29]:
# Check Merge
ords_prods_cust.shape

(32404859, 31)

In [30]:
ords_prods_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 31 columns):
 #   Column                  Dtype         
---  ------                  -----         
 0   product_id              int64         
 1   product_name            object        
 2   aisle_id                int64         
 3   department_id           int64         
 4   prices                  float64       
 5   order_id                int64         
 6   user_id                 int64         
 7   order_number            int64         
 8   order_dow               int64         
 9   order_hour_of_day       int64         
 10  days_since_prior_order  float64       
 11  First_order             bool          
 12  add_to_cart_order       int64         
 13  reordered               int64         
 14  _merge                  category      
 15  Busiest_days            object        
 16  Slowest_days            object        
 17  busiest_period_of_day   object        
 18  

I merged the cleaned customer dataset with my main Instacart data (ords_prods_merge) using the user_id column as the key.
Before merging, I confirmed that both columns had the same data type (int64).
I used a left join to keep all existing Instacart orders while adding customer information.
The resulting dataset (ords_prods_cust) now combines order, product, and customer data for analysis.

# 7.Ensure your notebook contains logical titles, section headings, and descriptive code comments.

# 8.Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [32]:
## Export data
ords_prods_cust.to_pickle(os.path.join(path, '02 - Data', 'prepared data', 'ords_prods_cust_task_4.9.pkl'))

In [33]:
## Export data 
ords_prods_merge.to_pickle(os.path.join(path,'02 - Data','prepared data','ords_prods_merge_task_4.8_Task.pkl'))

In [34]:
## Expored cleaned customers data
customers.to_pickle(os.path.join(path, '02 - Data', 'prepared data', 'customers_cleaned.pkl'))